harness/determined/core/_metrics.py

```diff
@@ -0,0 +1,207 @@
+        report_time: Optional[datetime.datetime] = None,
+    ) -> None:
+        """Adds metrics to a queue to be reported.
+
+        Before publishing to the queue, check for exceptions that might have occurred in the
+        background reporting thread from a previous report and raise them.
+        """
+        self._maybe_raise_exception()
+        self._shipper.publish_metrics(
+            group=group,
+            steps_completed=steps_completed,
+            metrics=metrics,
+            batch_metrics=batch_metrics,
+            report_time=report_time,
+        )
+
+    def _maybe_raise_exception(self) -> None:
+        """Check the error queue for exceptions and raise if there are any."""
+        if not self._error_queue.empty():
+            err_msg = self._error_queue.get(block=False)
+            logger.error(f"Error reporting metrics: {err_msg}")
+            raise err_msg
+
+    def start(self) -> None:
+        self._shipper.start()
+
+    def close(self) -> None:
+        self._maybe_raise_exception()
+        self._shipper.stop()
+        self._join_with_timeout()
+
+    def _join_with_timeout(self) -> None:
+        while not self._shipper._queue.empty():
+            # If the queue isn't empty, wait for metric reporting and print logs intermittently.
+            self._shipper.join(timeout=10)
+            self._maybe_raise_exception()
+            if self._shipper.is_alive():
+                logger.info("Waiting for _Shipper thread to finish reporting metrics...")
+            else:
+                return
+
+        # Metrics queue is empty, join with timeout to avoid hangs and add logs to help debug.
+        self._shipper.join(timeout=1)
+        if self._shipper.is_alive():
+            logger.info(f"Waiting for _Shipper thread to finish...")
+            self._shipper.join(timeout=5)
+            if self._shipper.is_alive():
+                logger.warning(f"Failed to complete _Shipper cleanup.")
+            else:
+                logger.info(f"_Shipper cleanup complete.")
+
+
+class _TrialMetrics:
+    def __init__(
+        self,
+        group: str,
+        steps_completed: int,
+        metrics: Dict[str, Any],
+        batch_metrics: Optional[List[Dict[str, Any]]] = None,
+        report_time: Optional[datetime.datetime] = None,
+    ):
+        self.group = group
+        self.steps_completed = steps_completed
+        self.metrics = metrics
+        self.batch_metrics = batch_metrics
+        self.report_time = report_time
+
+
+class _Shipper(threading.Thread):
+    METRICS_QUEUE_MAXSIZE = 1000
+
+    def __init__(
+        self,
+        session: api.Session,
+        trial_id: int,
+        run_id: int,
+        error_queue: queue.Queue,
+    ):
+        self._queue = queue.Queue(maxsize=self.METRICS_QUEUE_MAXSIZE)
+        self._error_queue = error_queue
+        self._session = session
+        self._trial_id = trial_id
+        self._run_id = run_id
+
+        super().__init__()
```

Comment:
lgtm, one last debugging convenience

```suggestion
        super().__init__(name="MetricsShipperThread")
```